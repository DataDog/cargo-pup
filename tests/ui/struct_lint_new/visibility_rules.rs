// This product includes software developed at Datadog (https://www.datadoghq.com/) Copyright 2024 Datadog, Inc.

//@compile-flags: --crate-name test_visibility_rules
//@compile-flags: --crate-type lib

// This test verifies that the StructRule::MustBePrivate and StructRule::MustBePublic rules work correctly

// Testing MustBePrivate rule:
pub struct InternalData { //~ ERROR: Struct 'InternalData' is public, but must be private
    field: i32,
}

// This should not trigger any warnings/errors (private, as required)
struct CorrectlyPrivate {
    value: String,
}

// Testing MustBePublic rule:
struct HiddenApi { //~ WARN: Struct 'HiddenApi' is private, but must be public
    id: u64,
}

// This should not trigger any warnings/errors (public, as required)
pub struct CorrectlyPublic {
    data: Vec<u8>,
}

// Also testing name-pattern + visibility combination
pub struct InternalModel { //~ ERROR: Struct 'InternalModel' is public, but must be private
    sensitive_data: String,
}

// This should not trigger any warnings/errors (private, as required by pattern match)
struct AnotherInternalModel {
    more_data: bool,
} 
