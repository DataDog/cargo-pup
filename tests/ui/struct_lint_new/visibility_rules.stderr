error: Struct 'InternalData' is public, but must be private
  --> tests/ui/struct_lint_new/visibility_rules.rs:9:1
   |
LL | pub struct InternalData {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: Remove the 'pub' visibility modifier
   = note: Applied by cargo-pup rule 'must_be_private_test'.
   = note: `#[deny(struct_lint_must_be_private)]` on by default

warning: Struct 'HiddenApi' is private, but must be public
  --> tests/ui/struct_lint_new/visibility_rules.rs:19:1
   |
LL | struct HiddenApi {
   | ^^^^^^^^^^^^^^^^^^
   |
   = help: Add the 'pub' visibility modifier
   = note: Applied by cargo-pup rule 'must_be_public_test'.
   = note: `#[warn(struct_lint_must_be_public)]` on by default

error: Struct 'InternalModel' is public, but must be private
  --> tests/ui/struct_lint_new/visibility_rules.rs:29:1
   |
LL | pub struct InternalModel {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: Remove the 'pub' visibility modifier
   = note: Applied by cargo-pup rule 'name_and_visibility_test'.

error: aborting due to 2 previous errors; 1 warning emitted

